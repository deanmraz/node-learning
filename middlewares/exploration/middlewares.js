
export const withMiddlewares = async (handler, middlewares = []) => {

  const wrapHandler = async (event, context) => {
    const Request = await handler(event, context);

    const chainMiddlewares = ([firstMiddleware, ...restOfMiddlewares]) => {
      if (firstMiddleware) {
        console.log('setup', firstMiddleware.toString());
        const wrapMiddleware = async (request) => {
          console.log('calling', firstMiddleware);
          return firstMiddleware(request, chainMiddlewares(restOfMiddlewares));
        };
        return wrapMiddleware;
      } else {
        console.log('no more middlewares to wrap');
        return 'END';
      }
    };

    if (middlewares.length > 0) {
      const wrappedMiddlewares = chainMiddlewares(middlewares);
      return wrappedMiddlewares(Request);
    }
    return Request;
  };
  return wrapHandler;
};
