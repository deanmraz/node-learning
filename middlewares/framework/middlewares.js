
const chainMiddlewares = ([firstMiddleware, ...restOfMiddlewares]) => {
  if (firstMiddleware) {
    return async request => firstMiddleware(request, chainMiddlewares(restOfMiddlewares));
  }
  // NOTE nothing us to run so return undefined
};

export const withMiddlewares = async (handler, middlewares = []) => {
  const wrapHandler = async (event, context) => {
    const Request = await handler(event, context);
    if (middlewares.length > 0) {
      return chainMiddlewares(middlewares)(Request);
    }
    return Request;
  };
  return wrapHandler;
};
