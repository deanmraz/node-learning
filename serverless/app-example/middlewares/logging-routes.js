export { default } from '@framework/middlewars/logging-routes';
