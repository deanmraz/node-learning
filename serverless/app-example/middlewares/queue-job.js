export default class Middleware {

  jobQueue: false;

  async handle(ctx, next) {

    if (this.request.queue) {
      // queue up request into sqs job

    }
    return await next();
  }

}
