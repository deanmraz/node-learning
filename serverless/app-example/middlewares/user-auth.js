export default class Middleware {

  jobQueue: false;

  constructor(UserAuth) {
    this.user = UserAuth;
  }

  async handle(ctx, next) {
    if (this.user.authenticated) {
      return await next();
    }
    //REVIEW throw validation error
  }

}
