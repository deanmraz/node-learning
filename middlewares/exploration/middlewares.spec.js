import { withMiddlewares } from './middlewares';

export const timeout = async (ms) => {
  return await new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve(true);
    }, ms);
  });
};

const runtimeMiddleware = async (request, next) => {
  const start = Date.now();
  // await timeout(1000);
  console.log(next);
  const response = await next(request);
  console.log(response);
  response.headers['run-time'] = Date.now() - start;
  return response;
};
const mainMiddleware = async (request, next) => {
  console.log('should be undefined', next);
  const { aws } = request;
  return {
    statusCode: 200,
    body: JSON.stringify(aws),
    headers: {
      'Content-Type': 'text/plain',
    },
  };
};

const AwsHandler = async (event, context) => {
  return {
    aws: {
      event,
      context,
    },
  };
};

describe('Middlewares', () => {
  test('Return Handler Request Data', async () => {
    const handler = await withMiddlewares(AwsHandler);
    const response = await handler({ event: 1 }, { context: 1 });
    expect(response).toMatchSnapshot();
  });
  test('Chain Middlewares', async () => {
    const handler = await withMiddlewares(AwsHandler, [runtimeMiddleware, mainMiddleware]);
    const response = await handler({ event: 1 }, { context: 1 });
    expect(response).toMatchSnapshot();
  });
});
