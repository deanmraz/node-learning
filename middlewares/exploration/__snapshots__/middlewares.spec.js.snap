// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Middlewares Chain Middlewares 1`] = `
Object {
  "body": "{\\"event\\":{\\"event\\":1},\\"context\\":{\\"context\\":1}}",
  "headers": Object {
    "Content-Type": "text/plain",
    "run-time": 9,
  },
  "statusCode": 200,
}
`;

exports[`Middlewares Return Handler Request Data 1`] = `
Object {
  "aws": Object {
    "context": Object {
      "context": 1,
    },
    "event": Object {
      "event": 1,
    },
  },
}
`;
